### PR TITLE
Modify windows package provider to allow url

### PR DESCRIPTION
**Pull request by jdmundrawala**
This patch allows people to specify a url for source for the `windows_package` resource.

cc @btm @smurawski @ksubrama

### CODE REVIEW

File: lib/chef/provider/package/windows.rb [hunk 1]
```diff
@@ -36,19 +37,23 @@ class Windows < Chef::Provider::Package
 
         # load_current_resource is run in Chef::Provider#run_action when not in whyrun_mode?
         def load_current_resource
-          @new_resource.source(Chef::Util::PathHelper.validate_path(@new_resource.source))
-
           @current_resource = Chef::Resource::WindowsPackage.new(@new_resource.name)
-          @current_resource.version(package_provider.installed_version)
-          @new_resource.version(package_provider.package_version)
-          @current_resource
+          if should_download?
```
Comment (btm) [inline lib/chef/provider/package/windows.rb]: So the obvious design debate here is, should we download the file during load_current_resource?

Comment (jdmundrawala) [inline lib/chef/provider/package/windows.rb]: What are the advantages of doing it during load_current_resource?

File: lib/chef/provider/package/windows.rb [hunk 2]
```diff
@@ -80,6 +93,63 @@ def install_package(name, version)
...
+              filename = ::File.basename(::URI.unescape(uri.path))
+              file_cache_dir = Chef::FileCache.create_cache_path("package/")
+              Chef::Util::PathHelper.cleanpath("#{file_cache_dir}/#{filename}")
+            else
+              Chef::Util::PathHelper.cleanpath(new_resource.source)
+            end
+          end
+        end
+
+        def is_url?(source)
```
Comment (btm) [inline lib/chef/provider/package/windows.rb]: I see this example code on the tubes, but it is also basically the same code as ```Chef::Provider::Package::Rpm.uri_scheme?```. I wonder if we have to hard code the list of schemes here? It'd be best if we could just have that in Chef::Provider::RemoteFile::Fetcher and let that raise the error.

URI.scheme_list seems limited, like ```URI.scheme_list.include?(URI.parse(source).scheme.upcase)```. Maybe a constant like ```Chef::Provider::RemoteFile::Fetcher::SUPPORTED_SCHEMES```? Maybe it doesn't matter.

File: lib/chef/provider/package/windows.rb (hunk 1 above)
Comment (btm) [inline lib/chef/provider/package/windows.rb]: During compile, without the file we don't know the product id (for MSI), so we don't know if the package is installed or not. Aren't we going to always decide that it's not installed and always run the install action during converge?

File: lib/chef/provider/package/windows.rb [hunk 3]
```diff
@@ -80,6 +93,63 @@ def install_package(name, version)
...
+            r.options(new_resource.options)
+          end
+        end
+
+        def download_source_file
+          source_resource.run_action(:create)
+          Chef::Log.debug("#{@new_resource} fetched source file to #{source_resource.path}")
+        end
+
+        def source_resource
```
Comment (btm) [inline lib/chef/provider/package/windows.rb]: how do you set checksum on the remote_file resource? Can we have an attribute that we pass a hash of attributes that gets set on the remote_file resource?

File: lib/chef/resource/windows_package.rb [hunk 4]
```diff
@@ -69,10 +69,26 @@ def source(arg=nil)
...
+            @source = arg
+          else
+            @source = ::File.absolute_path(arg).gsub(::File::SEPARATOR, ::File::ALT_SEPARATOR)
+          end
         end
       end
+
+      private
+
+      def is_url?(source)
```
Comment (btm) [inline lib/chef/resource/windows_package.rb]: Whatever we do, this makes three places we have this method (this resource/provider, plus rpm provider), it should probably be a mixin.

File: lib/chef/provider/package/windows.rb (hunk 1 above)
Comment (jdmundrawala) [inline lib/chef/provider/package/windows.rb]: Kind of... this resource will become 2 resources, and one of those 2 will need to be converged.

File: lib/chef/provider/package/windows.rb (hunk 3 above)
Comment (jdmundrawala) [inline lib/chef/provider/package/windows.rb]: Yep, that shouldn't be hard to do. However, as I went down this road, it sort of felt like if you wanted to do anything fancy like that, you should just use remote file. Not sure if that is a valid argument as I do feel like I'm a little biased here.

I originally wanted to let people pass a remote file resource as the source attribute (with all the chef dsl-ing), however at that point, you may as well just declare it with the dsl that's already there.

File: lib/chef/provider/package/windows.rb [hunk 5]
```diff
@@ -80,6 +93,63 @@ def install_package(name, version)
...
+        # @return [Array] new_version(s) as an array
+        def new_version_array
+          # Because the one in the parent caches things
+          [new_resource.version]
+        end
+
+        private
+
+        def should_download?
+          is_url?(new_resource.source) && !::File.exists?(source_location)
```
Comment (btm) [inline lib/chef/provider/package/windows.rb]: If we have a URL and an old version of the package on disk?

Comment (jdmundrawala) [inline lib/chef/provider/package/windows.rb]: yep, that's wrong. good catch

File: lib/chef/provider/package/windows.rb (hunk 2 above)
Comment (jdmundrawala) [inline lib/chef/provider/package/windows.rb]: I've mixed too many concepts here. `uri_scheme?` shouldn't care about what schemes are supported, seems like remote_file's job. I'm going to update it for that, let me know what you think.

**Comment by jdmundrawala** [comment]
@btm added mixin for Uris and a way to override remote_file attributes

**Comment by ksubrama** [comment]
Left shark gives a :+1:

**Comment by smurawski** [comment]
@jdmundrawala How do the remote file attributes get surfaced in the DSL?  Are they top level attributes with the standard package ones?  Or are they nested?  If nested, we should really surface at least checksum as a top-level attribute (to smooth the transition from the Windows cookbook).

File: lib/chef/mixin/uris.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'uri'
+
+class Chef
+  module Mixin
+    module Uris
+      def uri_scheme?(source)
```
Comment (btm) [inline lib/chef/mixin/uris.rb]: Can you add a comment here about what this returns? I'm guessing it's true if there is a scheme in the ```scheme://``` format?

File: lib/chef/provider/package/windows.rb [hunk 7]
```diff
@@ -80,6 +94,60 @@ def install_package(name, version)
...
+          source_resource.run_action(:create)
+          Chef::Log.debug("#{@new_resource} fetched source file to #{source_resource.path}")
+        end
+
+        def source_resource
+          @source_resource ||= Chef::Resource::RemoteFile.new(default_download_cache_path, run_context).tap do |r|
+            r.source(new_resource.source)
+            r.backup(false)
+
+            if new_resource.remote_file_attributes
```
Comment (btm) [inline lib/chef/provider/package/windows.rb]: Can we get a RELEASE_NOTES.md entry with an example of how to use this?

File: lib/chef/mixin/uris.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        # From open-uri
+        !!(%r{\A[A-Za-z][A-Za-z0-9+\-\.]*://} =~ source)
```
Comment (btm) [inline lib/chef/mixin/uris.rb]: This is nice. It looks like it works correctly in all the cases I was worried about before.

File: lib/chef/resource/windows_package.rb [hunk 9]
```diff
@@ -69,10 +71,22 @@ def source(arg=nil)
           @source
         else
           raise ArgumentError, "Bad type for WindowsPackage resource, use a String" unless arg.is_a?(String)
-          Chef::Log.debug("#{package_name}: sanitizing source path '#{arg}'")
-          @source = ::File.absolute_path(arg).gsub(::File::SEPARATOR, ::File::ALT_SEPARATOR)
+          if uri_scheme?(arg)
+            @source = arg
+          else
+            @source = ::File.absolute_path(arg).gsub(::File::SEPARATOR, ::File::ALT_SEPARATOR)
```
Comment (btm) [inline lib/chef/resource/windows_package.rb]: Probably should switch to ```Chef::Util::PathHelper.canonical_path``` now, which would also add long path support with the \\?\ business.

File: lib/chef/provider/package/windows.rb [hunk 10]
```diff
@@ -80,6 +94,60 @@ def install_package(name, version)
...
+
+        # @return [Array] new_version(s) as an array
+        def new_version_array
+          # Because the one in the parent caches things
+          [new_resource.version]
+        end
+
+        private
+
+        def download_file_missing?
```
Comment (btm) [inline lib/chef/provider/package/windows.rb]: downloadable_file_missing?

File: lib/chef/resource/windows_package.rb (hunk 9 above)
Comment (jdmundrawala) [inline lib/chef/resource/windows_package.rb]: msiexec does not like it

**Comment by jdmundrawala** [comment]
surfaced checksum attribute

File: lib/chef/resource/windows_package.rb (hunk 9 above)
Comment (btm) [inline lib/chef/resource/windows_package.rb]: The \\\\?\\ prepend? You can pass an argument to canonical_path to not add it.

Comment (jdmundrawala) [inline lib/chef/resource/windows_package.rb]: done

**Comment by smurawski** [comment]
:+1: Rock on Jay.

**Comment by btm** [comment]
:+1: :+1: 

Awesome work. /cc @jeremiahsnapp

**Pull request merged**